prov/util/src/util_buf.c

```diff
@@ -179,7 +179,10 @@ int ofi_bufpool_grow(struct ofi_bufpool *pool)
 	if (pool->attr.free_fn)
 	    pool->attr.free_fn(buf_region);
 err2:
-	ofi_freealign(buf_region->alloc_region);
+	if (pool->attr.flags & OFI_BUFPOOL_MMAPPED)
+		ofi_free_hugepage_buf(buf_region->alloc_region, pool->alloc_size);
+	else
+		ofi_freealign(buf_region->alloc_region);
```

Comment:
I've cherry picked this commit for v1.7.x, then changed every line except for 'else' so that it would work.  :)

Reply:
There's a deeper problem here.  The freeing of the allocated region needs to be based on the allocation method, not the current state of the pool flags.  While I think the above works in this error flow, a similar problem occurs in ofi_bufpool_destroy().  The buffer pool can end up with a mix of huge and normal (i.e. malloc) allocations.  We need to track this per buffer region.  I will prepare a fix for this.

Reply:
As it stands, you can't really get a mixed allocation within a pool right, since the allocs are also based off the attribute (and if the allocation failed the first time around, pool growth will not try to get huge pages again)?

Reply:
You're right, we only fall back if we haven't allocated anything.  So there's no bug to fix.  For some reason I was thinking we'd fall back to normal allocations if any huge page allocation failed, but that's not the case.